source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -0,0 +1,186 @@
+#include "extensions/filters/listener/tls_inspector/tls_inspector.h"
+
+#include <arpa/inet.h>
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/stats/stats.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/config/well_known_names.h"
+
+#include "openssl/bytestring.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace TlsInspector {
+
+namespace {
+
+// This could use the same index as Ssl::ContextImpl, but allocating 1 extra
+// index seems better than adding the coupling between components.
+int sslIndex() {
+  CONSTRUCT_ON_FIRST_USE(int, []() -> int {
+    int ssl_index = SSL_get_ex_new_index(0, nullptr, nullptr, nullptr, nullptr);
+    RELEASE_ASSERT(ssl_index >= 0);
+    return ssl_index;
+  }());
+}
+
+} // namespace
+
+Config::Config(Stats::Scope& scope)
+    : stats_{ALL_TLS_INSPECTOR_STATS(POOL_COUNTER_PREFIX(scope, "tls_inspector."))},
+      ssl_ctx_(SSL_CTX_new(TLS_with_buffers_method())) {
+  SSL_CTX_set_options(ssl_ctx_.get(), SSL_OP_NO_TICKET);
+  SSL_CTX_set_session_cache_mode(ssl_ctx_.get(), SSL_SESS_CACHE_OFF);
+  SSL_CTX_set_tlsext_servername_callback(
+      ssl_ctx_.get(), [](SSL* ssl, int* out_alert, void*) -> int {
+        Filter* filter = static_cast<Filter*>(SSL_get_ex_data(ssl, sslIndex()));
+        filter->onServername(SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name));
+
+        // Return an error to stop the handshake; we have what we wanted already.
+        *out_alert = SSL_AD_USER_CANCELLED;
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+      });
+}
+
+bssl::UniquePtr<SSL> Config::newSsl() { return bssl::UniquePtr<SSL>{SSL_new(ssl_ctx_.get())}; }
+
+Filter::Filter(const ConfigSharedPtr config) : Filter(config, TLS_DEFAULT_MAX_CLIENT_HELLO) {}
+
+Filter::Filter(const ConfigSharedPtr config, uint32_t max_client_hello_size)
+    : config_(config), ssl_(config_->newSsl()) {
+  SSL_set_ex_data(ssl_.get(), sslIndex(), this);
+  SSL_set_accept_state(ssl_.get());
+
+  // TODO(ggreenway) PERF: Put a buffer in thread-local storage and have every Filter share
+  // the buffer.
+  buf_.reserve(max_client_hello_size);
```

Comment:
This is done to avoid needing to zero-initialize the data. But yeah, it wouldn't surprise me if there's some checking for this. I'll change it; there's already a PERF comment here.